docs/posts/duckdb-for-rag/index.qmd

```diff
@@ -0,0 +1,543 @@
+
+```bash
+pip install 'ibis-framework[duckdb,examples]' openai marvin python-dotenv
+```
+
+Then import them:
+
+```{python}
+import ibis
+import marvin
+
+from openai import OpenAI
+
+from dotenv import load_dotenv
+```
+
+You need to set the `OPENAI_API_KEY` environment variable to your OpenAI API
+key. You can do this by creating a `.env` file in the same directory as your
+code and adding the following line:
+
+```
+OPENAI_API_KEY=<your-api-key>
+```
+
+Then, you can load the environment variables with the following code:
+
+```{python}
+load_dotenv()
+```
+
+Now, we're ready to make LLM calls. Let's setup Ibis for interactive use:
+
+```{python}
+ibis.options.interactive = True
+ibis.options.repr.interactive.max_rows = 20
+ibis.options.repr.interactive.max_columns = None
+```
+
+and setup Marvin to used the latest GPT-4-turbo model:
+
+```{python}
+marvin.settings.openai.chat.completions.model = "gpt-4-0125-preview"
+marvin.settings.openai.chat.completions.model
+```
+
+and get our penguins:
+
+```{python}
+penguins = ibis.examples.penguins.fetch()
+penguins.limit(3)
+```
+
+## Demonstration
+
+We'll start with a slightly modified text-to-SQL function [stolen from our
+previous "LMs for data" blog](../lms-for-data/index.qmd):
+
+```{python}
+@marvin.fn
+def _text_to_sql(
+    text: str,
+    table_name: str,
+    table_schema: str,
+    table_preview: str,
+) -> str:
+    """Writes a SQL SELECT statement for the `text` given the provided
+    `table_name`, `table_schema`, and `table_preview`.
+
+    Spell out all column names, never use SELECT *. Include ALL columns unless told otherwise.
+    Use newlines and indentation for readability. Use LIMIT, not TOP.
+    Select FROM the `table_name` only (penguins), you have no access to other tables.
+    """
+
+
+def text_to_sql(
+    text: str,
+    table_name: str,
+    table_schema: str,
+    table_preview: str,
+) -> str:
+    sql = _text_to_sql(text, table_name, table_schema, table_preview)
+    return sql.strip().strip(";")
+```
+
+This will take the user's English (or really, any language the LLM sufficiently
+knows) query and turn it into SQL.
+
+:::{.callout-info title="Isn't this already RAG?"}
+We have retrieved information (a table name, schema, and preview) augmented our
+model's input with that information, and generated a response.
+:::
+
+Our hypothetical service takes text as input, generates SQL, and runs it against
+the database on behalf of the user. We need some metric to evaluate the user's
+feedback on the generated SQL. This could be an explicit input from the user
+(rate it from 1 to 10, or simply upvote/downvote the response) or inferred based
+on user behavior (did the curse out the bot or ask it to modify the SQL after
+initial generation). We'll ignore this feedback component for now, but in
+practice this would be a crucial piece of the system.
+
+Additionally, we might bootstrap the system with a set of known English and SQL
```

Comment:
You haven't mentioned what bootstrap is up to here, I think a like to what bootstrapping is would be helpful here.